098-using-image-volumes-to-improve-extensibility-of-Strimzi-operands.md

```diff
@@ -0,0 +1,335 @@
+For example:
+
+```
+$ docker build -t <my-registry.io>/<organization>/<image>:<tag> .
+$ docker push <my-registry.io>/<organization>/<image>:<tag>
+```
+
+Once this container image is pushed into the registry, it can be used as a source for the image volume.
+The container image contains only the plugin (and its dependencies).
+It does not contain any operating system artifacts or Strimzi JARs.
+Its rebuild is therefore needed only when a new version of the plugin is released.
+
+It is users responsibility to ensure that the files in the container image will not cause any class path conflicts and break the operand.
+(This is the same already today with the pre-existing mechanisms for adding plugins.)
+
+User is also responsible for securing the container images with the plugins and for securing the container registry.
+When pulling images and using them as volumes, Kubernetes will apply the same checks as with the regular container images used by the operand.
+Using the container images with plugins from untrusted sources might constitute a risk.
+
+### Kafka Connect plugins (connectors, transformations, etc.)
+
+Kafka Connect plugins continue to be handled in their own separate way even with image volumes through a new section in the `KafkaConnect` custom resource.
+This is because they are the most commonly used type of Apache Kafka plugins and deserve the _improved_ user experience.
+But it has its own technical reasons as well:
+* The connector plugins are stored in a separately defined path defined by the `plugin.path` option in Apache Kafka.
+  They are normally not included in the regular class path as most other plugins.
+* Having a separate API allows us to mix the existing approaches (Kafka Connect Build or custom container image) with the connectors mounted from image volumes.
+  This might be useful for example when some plugins are not (yet) available as container images.
+
+The plugins will be configured in the new section in `.spec.plugins`.
+The structure of this section is identical to the existing `.spec.build.plugins` used to define the plugins that should be added through the Kafka Connect Build feature.
+But the artifact types will be different.
+
+The `.spec.plugins` section will contain a list of plugins.
+Each plugin will consist of one or more artifacts to allow compose the plugin from different libraries if needed.
+As of this proposal, the only artifact type will be `image`.
+It will let users specify the reference to the container image with the plugin that will be mounted and used.
+The image can be specified using a tag or using a digest.
+The advantage of the digest is that it exactly specifies the container image that will be used.
+The advantage of the tag is that you can always use the latest version of the plugin without updating the custom resource.
+But it will not be deterministic which exact version is used.
```

Comment:
I'm wondering whether using a tag would work in practice. Couldn't there be some strange behaviour if the underlying Jar file for a connector is swapped out half way through processing data?

Reply:
It works like the container image. It would not update on the fly, but when the Pod restarts. So it would be the same as when the connectors are inside the main container.